### DISCUSSION TITLE
How to add header to GridHandle

### DISCUSSION THREAD

**Discussion by ogbozoyan** [eclipse-birt/birt#1389] [Q&A]
I need dynamically populate Grid by DTO in Java, but also i need that if grid goes on new page
report:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<report xmlns="http://www.eclipse.org/birt/2005/design" version="3.2.23" id="1">
    <property name="createdBy">Eclipse BIRT Designer Version 4.12.0.v202211281949</property>
    <property name="units">in</property>
    <property name="iconFile">/templates/blank_report.gif</property>
    <property name="bidiLayoutOrientation">ltr</property>
    <property name="imageDPI">72</property>
    <data-sources>
        <script-data-source name="Data Source" id="78"/>
    </data-sources>
    <data-sets>
        <script-data-set name="PojoDataSet" id="79">
            <list-property name="resultSetHints">
                <structure>
                    <property name="position">1</property>
                    <property name="name">columnName</property>
                    <property name="dataType">string</property>
                </structure>
                <structure>
                    <property name="position">2</property>
                    <property name="name">columnResponsible</property>
                    <property name="dataType">string</property>
                </structure>
                <structure>
                    <property name="position">3</property>
                    <property name="name">columnDeadline</property>
                    <property name="dataType">string</property>
                </structure>
                <structure>
                    <property name="position">4</property>
                    <property name="name">columnMark</property>
                    <property name="dataType">string</property>
                </structure>
            </list-property>
            <list-property name="columnHints">
                <structure>
                    <property name="columnName">columnName</property>
                    <text-property name="displayName">Наименование мероприятия</text-property>
                </structure>
                <structure>
                    <property name="columnName">columnResponsible</property>
                    <text-property name="displayName" key="responsible">Отвественный</text-property>
                </structure>
                <structure>
                    <property name="columnName">columnDeadline</property>
                    <text-property name="displayName">Срок</text-property>
                </structure>
                <structure>
                    <property name="columnName">columnMark</property>
                    <text-property name="displayName">Отметка о выполнении</text-property>
                </structure>
            </list-property>
            <list-property name="parameters">
                <structure>
                    <property name="name">firstRow</property>
                    <property name="dataType">string</property>
                    <property name="position">1</property>
                    <expression name="defaultValue" type="constant">""</expression>
                    <property name="isInput">true</property>
                    <property name="isOutput">true</property>
                </structure>
                <structure>
                    <property name="name">secondRow</property>
                    <property name="dataType">string</property>
                    <property name="position">2</property>
                    <expression name="defaultValue" type="constant">""</expression>
                    <property name="isInput">true</property>
                    <property name="isOutput">true</property>
                </structure>
                <structure>
                    <property name="name">thirdRow</property>
                    <property name="dataType">string</property>
                    <property name="position">3</property>
                    <expression name="defaultValue" type="constant">""</expression>
                    <property name="isInput">true</property>
                    <property name="isOutput">true</property>
                </structure>
            </list-property>
            <structure name="cachedMetaData">
                <list-property name="resultSet">
                    <structure>
                        <property name="position">1</property>
                        <property name="name">columnName</property>
                        <property name="dataType">string</property>
                    </structure>
                    <structure>
                        <property name="position">2</property>
                        <property name="name">columnResponsible</property>
                        <property name="dataType">string</property>
                    </structure>
                    <structure>
                        <property name="position">3</property>
                        <property name="name">columnDeadline</property>
                        <property name="dataType">string</property>
                    </structure>
                    <structure>
                        <property name="position">4</property>
                        <property name="name">columnMark</property>
                        <property name="dataType">string</property>
                    </structure>
                </list-property>
            </structure>
            <property name="dataSource">Data Source</property>
        </script-data-set>
    </data-sets>
    <page-setup>
        <simple-master-page name="mainPage" id="2">
            <property name="type">a4</property>
            <property name="orientation">landscape</property>
            <property name="topMargin">2cm</property>
            <property name="leftMargin">3cm</property>
            <property name="bottomMargin">2cm</property>
            <property name="rightMargin">1.5cm</property>
            <property name="showHeaderOnFirst">false</property>
            <property name="showFooterOnLast">false</property>
        </simple-master-page>
    </page-setup>
    <body>
        <grid name="testGrid" id="119">
            <column id="120"/>
            <column id="121"/>
            <column id="122"/>
            <column id="123"/>
            <row id="124">
                <cell id="125">
                    <label id="143">
                        <text-property name="text">Наименование мероприятия</text-property>
                    </label>
                </cell>
                <cell id="126">
                    <label id="144">
                        <text-property name="text" key="responsible">Отвественный</text-property>
                    </label>
                </cell>
                <cell id="127">
                    <label id="145">
                        <text-property name="text">Срок</text-property>
                    </label>
                </cell>
                <cell id="128">
                    <label id="146">
                        <text-property name="text">Отметка о выполнении</text-property>
                    </label>
                </cell>
            </row>
            <row id="129">
                <cell id="130"/>
                <cell id="131"/>
                <cell id="132"/>
                <cell id="133"/>
            </row>
            <row id="134">
                <cell id="135"/>
                <cell id="136"/>
                <cell id="137"/>
                <cell id="138"/>
            </row>
        </grid>
    </body>
</report>
```

Pseudocode:
```java
ReportDesignHandle designHandle = birtUtils.getReportDesignFromRunnable(runnable);
            GridHandle testGrid = (GridHandle) birtUtils.getItemByName(designHandle, "testGrid");

            List<PlanGraphDTO> dtos = new ArrayList<>();
            for (int i = 0; i < 50; i++) {
                String iStr = String.valueOf(i);
                dtos.add(PlanGraphDTO.builder()
                        .deadline(iStr)
                        .date(iStr)
                        .responsible(iStr)
                        .name(iStr)
                        .build());
            }

            StringBuilder concatenatedData = new StringBuilder();
            for (PlanGraphDTO g : dtos) {
                RowHandle newRow = birtUtils.createNewRow(designHandle);

                birtUtils.createCellWithTextToRow(designHandle, newRow, g.getName());
                birtUtils.createCellWithTextToRow(designHandle, newRow, g.getDate());
                birtUtils.createCellWithTextToRow(designHandle, newRow, g.getDeadline());
                birtUtils.createCellWithTextToRow(designHandle, newRow, g.getResponsible());

                birtUtils.addRowToGrid(testGrid, newRow);
            }


            log.debug("Running PlanGraph");
            task.run();
```

Link to birtUtils:
https://github.com/ogbozoyan/Birt-utils-java/blob/main/src/main/java/org/birtutils/BirtUtils.java

**Comment by ogbozoyan** [comment]
maven
```xml
<!--  BIRT  -->
        <dependency>
            <groupId>com.innoventsolutions.birt.runtime</groupId>
            <artifactId>org.eclipse.birt.runtime_4.8.0-20180626</artifactId>
            <version>4.8.0</version>
        </dependency>
        <dependency>
            <groupId>org.eclipse.birt.runtime</groupId>
            <artifactId>viewservlets</artifactId>
            <version>4.5.0</version>
        </dependency>
        <dependency>
            <groupId>commons-io</groupId>
            <artifactId>commons-io</artifactId>
            <version>2.11.0</version>
        </dependency>
<!--  /BIRT  -->
```

**Comment by ogbozoyan** [comment]
Maybe there's some way how to populate dataset inside code  with my POJO and still use TableHandle instead of GridHandle

**Comment by speckyspooky** [comment]
The standard BIRT-grid doesn't includ header or footer. The grid-framing use internally and also at report side only row.
The header and footer row/object is only given on a real table-element (which require a dataset), therefore TableHandle make more sense.

I don't know why you will try to add the header dynamically. Perhaps anaother option for you would be to add the header row(s) by default and use the visibility to control the display of it with according script which checks your parameters or flags or global Java object status flags.

**Reply by ogbozoyan** [reply]
Could you please provide example of that script, i tried 3 days in row and i couldn't make it, even with chat gpt

**Reply by speckyspooky** [reply]
The question is under which conditions do you want that the header will be printed. Of course I could provide a very simple report which use a parameter to switch on/off the header rows.

**Reply by ogbozoyan** [reply]
Condition:
This is Grid with 4 columns and first row is "header", i'm will in loop add data and when grid continues on next page i need that first row on new page was header, like a table works

**Reply by speckyspooky** [reply]
But why you doesn't use a standard table with normal header. So it will repeated on each following page starting with the first page.

**Reply by ogbozoyan** [reply]
I didn't realise  and there's no guids how to add in table in Java code new rows
I found only this guide https://www.vogella.com/tutorials/EclipseBIRT/article.html#datasource-and-dataset ,this case good one, but i'm rendering my reports in java by   IRunAndRenderTask and i don't know how to add my List<PojoDTO> to this report and would it work how it needs for me. But i found perfect solution for population of grids by code. 
It would be better if the java code would show how to add dynamically rows instead of a script in birt

**Reply by ogbozoyan** [reply]
Im working on service that takes data from internal service and renders report based on that data,
workflow:
1. accepting json
2. processing, replacing parameters where necessary
3. task.run()
4. and returning multipart 
Maybe this information will help idk...

**Comment by speckyspooky** [comment]
My idea is a little bit changed. I think I understand your target to to it full integrated in Java.
My thinking would be:
1. create a standard BIRT-report with designer and your requirements.
2. transform your data (Json, xml, etc.) to a standard data structure in such structure if you want e.g. POJO, XML, database
3. Call the report execution with you prepared data and you will get the report

In my mind you will have the full control of the report and you avoid the challenge to generate the report completely code based (could be a challenge for all elements).

I use this way to create log files integrated with combination of ETL-data creation.

Perhaps the idea would we usable in you case too.

**Reply by ogbozoyan** [reply]
That's sounds good, but how report will get my POJO ? I don't understand how to provide pojo inside report

**Reply by speckyspooky** [reply]
Yes, I researched the same page. The target is that a object like POJO must be provided to the report context
and there you can analyse it with the scripted data set. The scripted data set will create the internal data table defined by columns and rows.

Instead to use a POJO you could also create a String-array and provide it to the report context. In this case based on 2-dimensional you would have such kind of the table-structure of rows and columns.

The steps would be:
1. create your data object "String[][] myDataArray = ...;"
3. get the application context based on configuration or task: "HashMap hm = configOrTask.getAppContext();"
4. prepare the application context: hm.put("myDataArray", myDataArray);
5. set the application context: "configOrTask.setAppContext( hm );"

The array should now reachable on the application context. The advantage of the String-array beside the simulated table structure wpuld be the implementation on our BIRT report because you can simulate the array with a standard JavaScript-array. And then you can switch into the report if the "myDataArray != null" use it if not then use the JavaScript-array like for testing or fallback mechanism.

**Reply by ogbozoyan** [reply]
What type of configOrTask ?
 org.eclipse.birt.report.engine.api.EngineConfig or i can use IRunAndRenderTask ? 
 
My step:
1. Create scripted DS  
2. Attach data set
3. Create string interpretation of my pojo 
4. by script panel in birt designer proceed myDataArray 

Did I get it right?

**Reply by speckyspooky** [reply]
"configOrTask" of "configOrTask.getAppContext();" means that you have more like one option (thefore OR) which is dependend on your implementation. If you use "IRunTask task = engine.createRunTask(runnable);" the you can use "task.getAppContext();"

So yes, both options could be work:  org.eclipse.birt.report.engine.api.EngineConfig & IRunAndRenderTask
but I would test it with "IRunAndRenderTask"

And yes, your list sounds good to me.

**Reply by ogbozoyan** [reply]
I'll try and let you know it works out.
Thank you very much

**Reply by ogbozoyan** [reply]
One more question, when hm.put("myDataArray", myDataArray) my dataset name should be myDataArray ? Or it  independent of each other. 
Should i add columns to dataset manually and then map it with my incoming myDataArray in script , or it will be automatically mapped 
If it's not hard, could you provide report example please, i'll use it like a template

**Reply by speckyspooky** [reply]
For the "myDataArray" (1. parameter) this is only the name and myDataArray (2. parameter) is the object.
And yes, I would define the columns of the scripted dataset before to have the correct data types opf the columns istead of full dynamic.
So you can map the row-columns based on your array with at the fetch-method of the scripted dataset.

**Reply by ogbozoyan** [reply]
> For the "myDataArray" (1. parameter) this is only the name and myDataArray (2. parameter) is the object.
> 
> And yes, I would define the columns of the scripted dataset before to have the correct data types opf the columns istead of full dynamic.
> 
> So you can map the row-columns based on your array with at the fetch-method of the scripted dataset.

Ok, how to get variable to proceed and fetch inside pf designer in script panel?

**Comment by speckyspooky** [comment]
3 things for the scripted data set:

1. you will need a global variable "customCounter" (name like you want) to count the loops and to have it for the compare if the array was looped till your end
2. in the onFetch() you will set the values from the array to the current row. Be aware the fetch is a kind of endless loop till the return value is "false" (false = loop end, true = loop again)
3. access on your array, if it is give at the report context you can use it directly like:
onFetch();

```

if ( customCounter < myDataArray.size()) {
  row["col01"] = myDataArray[customCounter][0];
  row["col02"] = myDataArray[customCounter][1];
  row["col03"] = myDataArray[customCounter][2];
  row["col04"] = myDataArray[customCounter][3];
 customerCount++;
 return true;
}

return false;
```

Here you will find an according example:
http://blog.deepakazad.com/2010/01/scripted-data-sets-in-birt.html

**Reply by ogbozoyan** [reply]
> 3 things for the scripted data set:
> 
> 
> 
> 1. you will need a global variable "customCounter" (name like you want) to count the loops and to have it for the compare if the array was looped till your end
> 
> 2. in the onFetch() you will set the values from the array to the current row. Be aware the fetch is a kind of endless loop till the return value is "false" (false = loop end, true = loop again)
> 
> 3. access on your array, if it is give at the report context you can use it directly like:
> 
> onFetch();
> 
> 
> 
> ```
> 
> 
> 
> if ( customCounter < myDataArray.size()) {
> 
>   row["col01"] = myDataArray[customCounter][0];
> 
>   row["col02"] = myDataArray[customCounter][1];
> 
>   row["col03"] = myDataArray[customCounter][2];
> 
>   row["col04"] = myDataArray[customCounter][3];
> 
>  customerCount++;
> 
>  return true;
> 
> }
> 
> 
> 
> return false;
> 
> ```
> 
> 
> 
> Here you will find an according example:
> 
> http://blog.deepakazad.com/2010/01/scripted-data-sets-in-birt.html

Thank, I'll try
